File: app/src/main/java/com/qiaoqiao/repository/AbstractDsSource.java
package com.qiaoqiao.repository;


import android.net.Uri;
import android.support.annotation.NonNull;

import com.google.android.gms.maps.model.LatLng;
import com.qiaoqiao.repository.backend.Google;
import com.qiaoqiao.repository.backend.ImageProvider;
import com.qiaoqiao.repository.backend.Wikipedia;
import com.qiaoqiao.repository.backend.model.wikipedia.LangLink;

import java.io.File;

public abstract class AbstractDsSource {
	private Google mGoogle;
	private com.qiaoqiao.repository.backend.Wikipedia mWikipedia;
	private ImageProvider mImageProvider;

	protected AbstractDsSource() {

	}

	protected AbstractDsSource(@NonNull ImageProvider imageProvider) {
		mImageProvider = imageProvider;
	}

	protected AbstractDsSource(@NonNull Google google) {
		mGoogle = google;
	}

	protected AbstractDsSource(@NonNull Wikipedia wikipedia) {
		mWikipedia = wikipedia;
	}


	protected AbstractDsSource(@NonNull Google google, @NonNull Wikipedia wikipedia) {
		mGoogle = google;
		mWikipedia = wikipedia;
	}

	protected Google getGoogle() {
		return mGoogle;
	}

	protected Wikipedia getWikipedia() {
		return mWikipedia;
	}


	public ImageProvider getImageProvider() {
		return mImageProvider;
	}

	public void onBytes(@NonNull byte[] bytes, @NonNull DsLoadedCallback callback) {
	}

	public void onFile(@NonNull File file, @NonNull DsLoadedCallback callback) {
	}

	public void onUri(@NonNull Uri uri, @NonNull DsLoadedCallback callback) {
	}

	public void onKnowledgeQuery(int pageId, @NonNull DsLoadedCallback callback) {

	}

	public void onKnowledgeQuery(@NonNull String keyword, @NonNull DsLoadedCallback callback) {

	}

	public void onKnowledgeQuery(@NonNull LangLink langLink, @NonNull DsLoadedCallback callback) {

	}

	public void onGeosearchQuery(@NonNull LatLng latLng, long radius, @NonNull DsLoadedCallback callback) {

	}


	public void onTranslate(@NonNull String q, @NonNull DsLoadedCallback callback) {

	}

	public void onRecentRequest(@NonNull DsLoadedCallback callback) {

	}

	public void onImage(@NonNull DsLoadedCallback callback) {

	}
}

File: app/src/main/java/com/qiaoqiao/repository/AbstractDsSource.kt
package com.qiaoqiao.repository

import android.content.Context
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.vision.barcode.Barcode
import com.qiaoqiao.app.App
import com.qiaoqiao.repository.backend.Google
import com.qiaoqiao.repository.backend.ImageProvider
import com.qiaoqiao.repository.backend.ProductsService
import com.qiaoqiao.repository.backend.Wikipedia
import com.qiaoqiao.repository.backend.model.wikipedia.LangLink
import com.qiaoqiao.repository.database.LastLaunchImage
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.disposables.Disposable

abstract class AbstractDsSource() {
    protected var google: Google? = null
    protected var wikipedia: com.qiaoqiao.repository.backend.Wikipedia? = null
    protected var imageProvider: ImageProvider? = null
    protected var productsService: ProductsService? = null

    protected constructor(google: Google) : this() {
        this.google = google
    }

    protected constructor(imageProvider: ImageProvider) : this() {
        this.imageProvider = imageProvider
    }

    protected constructor(google: Google, wikipedia: Wikipedia, productsService: ProductsService) : this() {
        this.google = google
        this.wikipedia = wikipedia
        this.productsService = productsService
    }

    fun onLoadedLaunchImage(imageData: ByteArray, callback: DsLoadedCallback): Disposable {
        val flowable: Flowable<ByteArray> = Flowable.create({ emitter ->
            App.realm.executeTransaction {
                it.delete(LastLaunchImage::class.java)
                val lastLaunchImage = it.createObject(LastLaunchImage::class.java)
                lastLaunchImage.byteArray = imageData
                emitter.onNext(imageData)
            }
        }, BackpressureStrategy.BUFFER)
        return flowable.subscribe({
            callback.onSomeThingSuccessfully()
        }, { callback.onSomeThingUnsuccessfully() })
    }

    open fun onBytes(bytes: ByteArray, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onKnowledgeQuery(pageId: Int, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onKnowledgeQuery(keyword: String, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onKnowledgeQuery(langLink: LangLink, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onKnowledgeQuery(barcode: Barcode, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onGeosearchQuery(latLng: LatLng, radius: Long, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onTranslate(q: String, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onRecentRequest(callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }

    open fun onImage(cxt: Context, callback: DsLoadedCallback): Disposable {
        throw NotImplementedError()
    }
}